oak_client/src/attestation.rs

```diff
@@ -0,0 +1,153 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::{anyhow, Context};
+use http::uri::Uri;
+use tokio_rustls::rustls;
+use tonic::transport::{Channel, ClientTlsConfig};
+use x509_parser::{
+    der_parser::{self, der::parse_der, oid::Oid},
+    parse_x509_certificate,
+};
+
+// Using [`Nid::NETSCAPE_COMMENT`] identifier since `rust-openssl` doesn't support custom
+// extensions yet.
+// https://github.com/sfackler/rust-openssl/issues/1411
+// https://www.alvestrand.no/objectid/2.16.840.1.113730.1.13.html
+const TEE_EXTENSION_OID: Oid<'static> = der_parser::oid!(2.16.840.1.113730.1.13);
+
+/// Creates a remotely attested TLS channel for connecting to Oak.
+pub async fn create_attested_tls_channel(
+    uri: &Uri,
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<Channel> {
+    let tls_config = create_tls_attestation_config(root_tls_certificate, tee_quote)
+        .context("Couldn't create TLS config")?;
+    Channel::builder(uri.clone())
+        .tls_config(tls_config)
+        .context("Couldn't create TLS configuration")?
+        .connect()
+        .await
+        .context("Couldn't connect to Oak Application")
+}
+
+/// Create client TLS configuration with a custom X.509 certificate verifier.
+pub fn create_tls_attestation_config(
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<ClientTlsConfig> {
+    let mut config = rustls::ClientConfig::new();
+
+    // Enable HTTP/2.
+    const ALPN_H2: &str = "h2";
+    config.set_protocols(&[Vec::from(&ALPN_H2[..])]);
+
+    // Add root TLS certificate.
+    let mut cc_reader = std::io::BufReader::new(&root_tls_certificate[..]);
+    let certs = rustls::internal::pemfile::certs(&mut cc_reader)
+        .map_err(|error| anyhow!("Couldn't parse TLS certificate: {:?}", error))?;
+    if certs.len() > 1 {
```

Comment:
Could `certs` be empty or is that a case that `rustls::internal::pemfile::certs` handles as error?

Reply:
Updated to `!=`